Feature 28/pull gfs data

michelleharrold opened the pull request:
## Expected Differences ##

- [ ] Do these changes modify the system output in any way? **[Yes or No]**</br>
If **yes**, please describe:</br>

## Pull Request Testing ##

- [x] Describe testing already performed for these changes:</br>

On Cheyenne, I ran the run.py script for several initialization start times using the Caribbean-based namelist. Tests included changing the initialization time to test functionality to pull from NOMADS or S3 based on data availability. I also modified the input_freq_sec to test different input frequencies as well as testing 00 and 12 UTC initializations.

To pull data from NOMADS or S3, I modified the yml file to have an empty string for local_data.

Note 1: I did have to add several Python packages to the wrfcloud conda environment (e.g., boto3).

Note 2: Additional warning and error checking will need to be implemented.

- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>
@hahnd - Can you please do a review of the code changes to offer any Python-related suggestions as needed?
@fossell - Can you please run a couple new tests not using local_data? Running a time available on NOMADS as well as one not available would exercise the new functionality. The retention period of NOMADS is ~10 days.

To run (assuming your experiment is called 'test'):

`git clone -b feature-28/pull_gfs_data https://github.com/NCAR/wrfcloud`

```
cd $PATH_TO_CLONED_REPO/wrfcloud/python/src/wrfcloud/runtime
ln -sf /glade/p/ral/jntp/CCproject/wrfcloud/WRFV4.4_intel_dmpar/ WRFV4
ln -sf /glade/p/ral/jntp/CCproject/wrfcloud/WPSV4.4_intel_serial_large-file/ WPSV4
```

`cp /glade/p/ral/jntp/CCproject/wrfcloud/wpsprd/geo_em.d01.nc $PATH_TO_CLONED_REPO/wrfcloud/python/src/wrfcloud/runtime/domains/test`

Set up your test.yml in $PATH_TO_CLONED_REPO/wrfcloud/python/src/wrfcloud/runtime to have correct dates, intervals, and set local_data to be an empty string

Set up your namelist.wps in $PATH_TO_CLONED_REPO/wrfcloud/python/src/wrfcloud/runtime/domains/test

In $PATH_TO_CLONED_REPO/wrfcloud/python/src/wrfcloud/runtime:
`python run.py --name test `

- [ ] Do these changes include sufficient documentation updates, ensuring that no errors or warnings exist in the build of the documentation? **[No]**
No documentation was added for these mods. There should likely be some high-level description of where data is pulled from, if not provided locally.

- [ ] Do these changes include sufficient testing updates? **[Yes or No]**

- [ ] Will this PR result in changes to the test suite? **[Yes or No]**</br>
If **yes**, describe the new output and/or changes to the existing output:</br>

- [ ] Please complete this pull request review by **[12 August 2022]**.</br>

## Pull Request Checklist ##
- [x] Review the source issue metadata (labels, project, and milestone).
- [x] Complete the PR definition above.
- [x] Ensure the PR title matches the feature or bugfix branch name.
- [x] Define the PR metadata, as permissions allow.
Select: **Reviewer(s)**
Select: **Project**
Select: **Milestone** as the version that will include these changes
Select: **Development** to link to the original development issue.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue and delete your feature or bugfix branch from GitHub.

hahnd commented on python/src/wrfcloud/runtime/ungrib.py:
```diff
@@ -49,9 +60,74 @@ def get_grib_input(runinfo: RunInfo, logger: Logger) -> None:
+        cycle_dt_h = math.ceil(cycle_dt_s / 3600.)
+
+        # Set base URLs for NOMADS and S3 bucket with GFS data.
+        nomads_base_url = 'https://nomads.ncep.noaa.gov/pub/data/nccf/com/gfs/prod'
```
Consider moving these URLs to a configuration file rather than hard-coding here.  Perhaps wrfcloud/resources/env_vars.yaml would be appropriate, then you can get the values from `os.environ['name']`.

This would also allow us to set the URLs differently for different environments, such as running unit tests vs production.

hahnd commented on python/src/wrfcloud/runtime/ungrib.py:
```diff
@@ -49,9 +60,74 @@ def get_grib_input(runinfo: RunInfo, logger: Logger) -> None:
+
+            try:
+                full_url = os.path.join(nomads_base_url, gfs_file)
+                status_nomads = requests.get(full_url)
```
The `requests` and `urllib` libraries basically do the same thing, so I am surprised to see both of them used here.  I think you might be downloading the files twice.  With `requests.get`, you are downloading the data and only checking the HTTP response code, and then if that was successful, downloading again with `urllib.request.urlretrieve`.  Instead of calling `urllib.request.urlretrieve`, you should be able to save the data to a file with:

```
with gfs_file_out as open(gfs_file, 'wb'):
    gfs_file_out.write(status_nomads.content)
    gfs_file_out.close()
```

And maybe change the variable name of `status_nomads` to just `nomads` since it is not just the status.

Same duplicate-download comment goes for downloading the files from AWS.

hahnd commented on python/src/wrfcloud/runtime/ungrib.py:
```diff
@@ -4,9 +4,17 @@
 import itertools
+import math
+import os
+import requests
```
The `requests` library is a new requirement for the project, so it should be listed in the `install_requires` section of setup.py.  e.g.:

    install_requires=[
        ...
        'requests>=2.0'
    ],

hahnd reviewed (commented):
@michelleharrold - I agree with they way you implemented the downloads from S3.  As long as the URLs are public, we can just treat them as URLs and use the code to download either.  

The only reason to use boto3 would be if we were granted special permissions in their AWS account, or if they configured their bucket as requester-pays.

fossell reviewed (commented):
I was able to checkout this feature branch and run the test case with a) pulling from NOMADS with dates within the 10 day range, and b) pulling from S3 with dates outside of the 10 day range. Both completed successfully.  

Note - the testing instructions said to set up the namelist.wps in domains/test. I'm not sure what set up was needed and did not make any edits to it since relevant dates and whatnot should be made by the make_wps_namelist module. 

I did officially approve yet as I see David has requested some mods, so I can test again after those are made.

fossell commented on python/src/wrfcloud/runtime/ungrib.py:
```diff
@@ -49,9 +60,74 @@ def get_grib_input(runinfo: RunInfo, logger: Logger) -> None:
     else:
-        logger.debug('Getting GRIB file(s) from S3 bucket')
-        logger.warning('Not yet implemented!')
+        logger.debug('Getting GRIB file(s) from external source (NOMADS or S3 bucket)')
```
I guess just a general question about organization. I'm still wondering if get_grib_input should be in ungrib.py or be its own module, maybe in tools/.  In terms of functionality it doesn't really matter, ungrib needs the grib data to run.  But it's also sort of separate from ungrib too. I don't know, probably overthinking it.

michelleharrold replied:
@fossell - I think you and I had some discussion on this and decided to put it in the ungrib.py script. This was before we had planned on potentially including verification -- which would require pulling data. Given that we may need to pull data for other components and it is not part of running ungrib, I concur it might be better as its own module. Should I go ahead and do that?

mkavulich commented on python/src/wrfcloud/runtime/ungrib.py:
```diff
@@ -49,9 +60,74 @@ def get_grib_input(runinfo: RunInfo, logger: Logger) -> None:
             logger.debug(f'Linking input GRIB file {gribfile} to {griblink}')
             os.symlink(gribfile, griblink)
```
I know this was originally my code, but in retrospect we should move lines 57-61 to *after* this `if` block to avoid code redundancy.

michelleharrold replied:
@mkavulich - ah yes, good suggestion. Addressed in latest commits!

michelleharrold commented:
@hahnd, @fossell, @mkavulich -- I made a number of mods based on the helpful comments and suggestions above:

1) I pulled out the remote grib data acquisition and put it in tools/get_grib_input.py.
2) Moved NOMADS and AWS URLs to resources/env_var.yaml, so they aren't hardcoded (environment vars are initialized in run.py).
3) Removed duplicate download issue.
4) Some other code clean-up (e.g., only one section of code for linking) 

fossell reviewed (approved):
I re-tested both options: 1) pulling from S3 and 2) pulling from NOMADS, using the 10 day threshold for dates to test. Data pulled successfully and ungrib and metgrib ran successfully.

hahnd commented:
@michelleharrold - GitHub wasn't letting me make comments on the PR, so I put some comments into a new branch: origin/feature-28/pull_gfs_data_codereview_hahnd.

mkavulich commented:
Here are the changes from David's branch, in case anyone else was having trouble finding them: https://github.com/NCAR/wrfcloud/commit/671c4b48d450960da15780277025f9dd7f03bc2e

mkavulich commented:
@hahnd If we pull in these changes will the PR be okay to merge?

hahnd commented:
Yes, I think it is ok to merge.

michelleharrold commented:
I pulled in all of David's suggestion (sorry about all the typos, yikes!). I reran a couple quick tests. All is good.
